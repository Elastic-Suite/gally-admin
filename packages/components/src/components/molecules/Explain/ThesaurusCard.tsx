import React from 'react'
import { useRouter } from 'next/router'
import { styled } from '@mui/system'
import Button from '../../atoms/buttons/Button'
import Chip from '../../atoms/Chip/Chip'
import { TypographyOptions } from '@mui/material/styles/createTypography'

const ThesaurusCardStyled = styled('div')(({ theme }) => ({
  display: 'flex',
  flexDirection: 'column',
  justifyContent: 'center',
  alignItems: 'flex-start',
  gap: theme.spacing(3),
  background: theme.palette.colors.white,
  border: `1px solid ${theme.palette.colors.neutral[300]}`,
  padding: theme.spacing(4),
  '& h4': {
    ...(theme.typography as TypographyOptions).h4,
    color: theme.palette.colors.neutral[900],
    fontFamily: 'var(--gally-font)',
    margin: 0,
  },
}))

const TermsStyled = styled('div')(({ theme }) => ({
  display: 'flex',
  alignItems: 'flex-start',
  flex: 1,
  flexWrap: 'wrap',
  gap: theme.spacing(1),
}))

interface IProps {
  button: {
    label: string
    url: string
  }
  terms: string[]
  title: string
}

export default function ThesaurusCard({
  button,
  terms,
  title,
}: IProps): JSX.Element {
  const router = useRouter()

  function handleClick(): Promise<boolean> {
    return router.push(button.url)
  }

  return (
    <ThesaurusCardStyled>
      <h4>{title}</h4>
      <TermsStyled>
        {terms.map((term: string) => (
          <Chip key={term} color="primary" label={term} />
        ))}
      </TermsStyled>

      <Button display="secondary" onClick={handleClick} size="large">
        {button.label}
      </Button>
    </ThesaurusCardStyled>
  )
}
