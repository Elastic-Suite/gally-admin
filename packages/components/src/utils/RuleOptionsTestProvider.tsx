import React, { ReactNode } from 'react'
import {
  IRuleEngineOperators,
  RuleAttributeType,
  getIri,
} from '@elastic-suite/gally-admin-shared'

import ruleEngineOperator from '../../public/mocks/rule_engine_operators.json'
import sourceFields from '../../public/mocks/source_fields.json'
import sourceFieldLabels from '../../public/mocks/source_field_labels.json'

import RuleOptionsProvider from '../components/stateful-providers/RuleOptionsProvider/RuleOptionsProvider'

interface IProps {
  children: ReactNode
}

const sourceFieldLabelsMap = new Map(
  sourceFieldLabels['hydra:member'].map((label) => [
    label.sourceField,
    label.label,
  ])
)

const fields = sourceFields['hydra:member'].map((field) => {
  const label = sourceFieldLabelsMap.get(getIri('source_fields', field.id))
  return {
    id: field.id,
    code: field.code,
    label: (label || field.defaultLabel) ?? '',
    type: field.type as RuleAttributeType,
  }
})

function RuleOptionsTestProvider(props: IProps): JSX.Element {
  const { children } = props
  return (
    <RuleOptionsProvider
      catalogId={-1}
      localizedCatalogId={-1}
      fields={fields}
      ruleOperators={ruleEngineOperator as IRuleEngineOperators}
    >
      {children}
    </RuleOptionsProvider>
  )
}

export default RuleOptionsTestProvider
