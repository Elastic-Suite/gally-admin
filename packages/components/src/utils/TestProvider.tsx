import React, { ReactNode } from 'react'

import catalogs from '../../public/mocks/catalog.json'
import metadata from '../../public/mocks/metadata.json'

import { catalogContext, metadataContext } from '../contexts'
import { useCatalogs } from '../hooks'

import OptionsProvider from '../components/stateful-providers/OptionsProvider/OptionsProvider'

interface IProps {
  children: ReactNode
}

function TestProvider(props: IProps): JSX.Element {
  const { children } = props
  const contextValueCatalog = useCatalogs(catalogs['hydra:member'])
  const contextValueMetadata = metadata['hydra:member']

  return (
    <OptionsProvider>
      <catalogContext.Provider value={contextValueCatalog}>
        <metadataContext.Provider value={contextValueMetadata}>
          {children}
        </metadataContext.Provider>
      </catalogContext.Provider>
    </OptionsProvider>
  )
}

export default TestProvider
