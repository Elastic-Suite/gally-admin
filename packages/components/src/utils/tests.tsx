import React, { ReactNode } from 'react'
import { RenderOptions, render, renderHook } from '@testing-library/react'
import { PreloadedState } from '@reduxjs/toolkit'
import { Bundle, api } from '@elastic-suite/gally-admin-shared'

import { AppStore, RootState, setupStore } from '../store'

import AppProvider from '../components/stateful-providers/AppProvider/AppProvider'

import TestProvider from './TestProvider'

export interface IExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<RootState>
  store?: AppStore
}

export interface IRenderOutput {
  store: AppStore
}

const defaultState = {
  data: {
    api,
    bundles: [Bundle.VIRTUAL_CATEGORY],
    configurations: {
      'base_url/media': 'https://localhost/media/catalog/product/',
    },
    metadata: [
      {
        '@id': '/metadata/3',
        '@type': 'Metadata',
        id: 3,
        entity: 'product',
        sourceFields: ['/source_fields/217', '/source_fields/253'],
      },
      {
        '@id': '/metadata/4',
        '@type': 'Metadata',
        id: 4,
        entity: 'category',
        sourceFields: ['/source_fields/254', '/source_fields/286'],
      },
    ],
  },
  user: {
    requestedPath: '',
  },
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types, @typescript-eslint/explicit-function-return-type
export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = defaultState,
    store,
    ...renderOptions
  }: IExtendedRenderOptions = {}
) {
  store = store ?? setupStore(preloadedState)
  function Wrapper({ children }: { children: ReactNode }): JSX.Element {
    return (
      <AppProvider store={store}>
        <TestProvider>{children}</TestProvider>
      </AppProvider>
    )
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) }
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types, @typescript-eslint/explicit-function-return-type
export function renderHookWithProviders<R>(
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  hook: (props: any) => R,
  {
    preloadedState = defaultState,
    store,
    ...renderOptions
  }: IExtendedRenderOptions = {}
) {
  store = store ?? setupStore(preloadedState)
  function Wrapper({ children }: { children: ReactNode }): JSX.Element {
    return (
      <AppProvider store={store}>
        <TestProvider>{children}</TestProvider>
      </AppProvider>
    )
  }
  return { store, ...renderHook(hook, { wrapper: Wrapper, ...renderOptions }) }
}
