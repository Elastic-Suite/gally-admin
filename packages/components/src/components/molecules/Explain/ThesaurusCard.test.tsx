import React from 'react'
import { renderWithProviders } from '../../../utils/tests'
import ThesaurusCard from './ThesaurusCard'
import Router from 'next/router'
import { fireEvent, screen } from '@testing-library/react'

const args = {
  button: {
    label: 'Modify Thesaurus',
    url: '/admin/search/thesaurus/grid',
  },
  terms: ['Top', 'Jacket', 'Tank'],
  title: 'Synonyms and expansions',
}

describe('ThesaurusCard', () => {
  it('Should match snapshot', () => {
    const { container } = renderWithProviders(<ThesaurusCard {...args} />)

    expect(container).toMatchSnapshot()
  })

  it('Should push to the correct URL when the button is clicked', () => {
    const pushSpy = Router.push as jest.Mock
    pushSpy.mockClear()
    renderWithProviders(<ThesaurusCard {...args} />)
    const button = screen.getByRole('button')
    fireEvent.click(button)

    expect(pushSpy).toHaveBeenCalledWith(args.button.url)
  })
})
