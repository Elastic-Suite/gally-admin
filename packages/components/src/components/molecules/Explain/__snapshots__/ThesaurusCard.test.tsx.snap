// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThesaurusCard matches snapshot 1`] = `
<div>
  <div
    class="css-4ldiph"
  >
    <h4>
      Synonyms and expansions
    </h4>
    <div
      class="css-ts3zi"
    >
      <div
        class="MuiChip-root MuiChip-filled MuiChip-sizeMedium MuiChip-colorPrimary MuiChip-filledPrimary css-wih3ly-MuiChip-root"
        style="opacity: 1;"
      >
        <span
          class="MuiChip-label MuiChip-labelMedium css-6od3lo-MuiChip-label"
        >
          Top
        </span>
      </div>
      <div
        class="MuiChip-root MuiChip-filled MuiChip-sizeMedium MuiChip-colorPrimary MuiChip-filledPrimary css-wih3ly-MuiChip-root"
        style="opacity: 1;"
      >
        <span
          class="MuiChip-label MuiChip-labelMedium css-6od3lo-MuiChip-label"
        >
          Jacket
        </span>
      </div>
      <div
        class="MuiChip-root MuiChip-filled MuiChip-sizeMedium MuiChip-colorPrimary MuiChip-filledPrimary css-wih3ly-MuiChip-root"
        style="opacity: 1;"
      >
        <span
          class="MuiChip-label MuiChip-labelMedium css-6od3lo-MuiChip-label"
        >
          Tank
        </span>
      </div>
    </div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge css-156z0p6-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="button"
    >
      Modify Thesaurus
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;
